policies/codingstyle.txt

```diff
@@ -289,12 +289,69 @@ options.
+EVP_PKEY, X509 etc). When adding new API functions relevant to a given type, the
+following argument ordering should be used:
+
+    rettype TYPE_get_something(TYPE *,[args])
+    rettype TYPE_do_something(TYPE *,[args])
+    TYPE *TYPE_new([args])
+
+Where present the TYPE argument should always be first. In constructors or
+object factory type functions (such as a "fetch" function), the key argument
+that identifies the thing be constructed should be placed first (if such an
+argument exists). For example in a "fetch" function the name of the algorithm
+being fetched would be first.
+
+The remaining argument list should be put into order of importance. Deciding the
+relative importance of arguments is necessarily somewhat subjective, but the
+following guidelines should be considered:
+
+1) Any "context" arguments should be first. If multiple contexts are necessary
+   then they should be in order of most general to most specific. In some cases
+   (typically with an OPENSSL_CTX), a context may be NULL to indicate a default
+   context should be used. This does not lessen its importance. Since such a
+   context is always used, even though a default value may be applied, it does
+   not make it an "optional" argument.
+2) Required arguments should be placed after the context.
+3) Optional arguments (i.e. those that are only be used in certain circumstances
+   and therefore may be "NULL" or empty) are placed after required arguments.
+4) Output parameters are placed after optional arguments.
+5) Callbacks are placed after output parameters.
+
+Consistency with widely established practices may override these guidelines in
+some cases. For example in a "copy" type function, such as EVP_MD_CTX_copy, it
+is common practice to place the "destination" (an output parameter) first,
+followed by the "source" (a required input parameter).
+
 In function prototypes, include parameter names with their data types.
 Although this is not required by the C language, it is preferred in OpenSSL
 because it is a simple way to add valuable information for the reader.
 The name in the prototype declaration should match the name in the function
 definition.
 
+                Chapter 6.2: Extending existing functions
+
+From time to time it is necessary to extend an existing function. Typically this
+will mean adding additional arguments, but it may also include removal of some.
+
+Where an extended function should be added the orginal function should be kept
+and a new version created with the same name and an "_ex" suffix. For example,
+the "RAND_bytes" function has an extended form called "RAND_bytes_ex".
+
+Where an extended version of a function already exists and a second extended
+version needs to be created then it should have an "_ex2" suffix, and so on for
+further extensions.
+
+When an extended version of a function is created the order of existing
+parameters from the original function should be retained even where it does not
```

Comment:
Why?

It's a new function in the API, why not design it according to the guidelines?

The order won't break anyone's code, and even for the laziest of us (me) it does not really make a difference if I type the new arguments at the beginning, middle or end of the code change where I replace foo() with foo_ex().

Reply:
But this talks about order of existing arguments not about where you put the new ones. I actually think the proposal is right that we should not do reordering with the _ex functions as that is the most confusing thing. Removing unused params is definitely OK and although the least confusing thing would be also to put the new arguments at the end I can accept that sometimes for the clarity of the API call it is acceptable to put it somewhere else. But please no reordering of the old arguments.

Reply:
We must *not* change the *order* of existing parameters. This would be very confusing.

Reply:
So if we cannot change the order of existing parameters, how can we go from the point in which our API is inconsistent to a point in a future release in which the API will not be inconsistent because of legacy?

I really keep failing the "confusion cost" for our userbase: they are already editing their source code to replace `foo(a, out, in)` to instead be `foo_ex(...)`.
Let's assume that we need an extra argument `barctx`, and that with proper guidelines in place we are trying to reach a point where `output` parameters should always be after `input` parameters.
Why is `foo_ex(a, in, out, barctx)` more confusing than `foo_ex(a, out, in, barctx)` when clearly the latter is the confusing one in the scope of our updated API where we strive for `output`s after `input`s?

I don't think the confusion stems from having to reorder the arguments, but from not being consistent in how to order stuff across different portions of our public API.

I would actually argue that it is more confusing not to reorder the arguments of an extended existing function if the existing sorting is incompatible with our formalized policy:
- a new function is added to the API that intentionally breaks our argument sorting policy, so you can't trust new functions to do the right thing, but you have to constantly check which ordering is being used
- as time goes by consistency keeps decreasing, because we keep mixing consistently sorted new stuff with inconsistenly sorted new extended functions that propagate the quirkiness due to legacy
- we are left without a incremental way to steer the boat towards a unified and consistent API design

Reply:
>> We must not change the order of existing parameters. This would be very confusing.

> Why is foo_ex(a, in, out, barctx) more confusing than foo_ex(a, out, in, barctx) when clearly the latter is the confusing one in the scope of our updated API where we strive for outputs after inputs?

I tend towards @mattcaswell's point of view, because I think that reordering the function parameters while extending the parameter list can be error prone and introduce runtime errors. For example, if we swap parameters of the same type while adding others,

    -int foo(FOO_CTX *a, char *out, char *in);
	+int foo_ex(FOO_CTX *a, char *in, char *out, BAR_CTX *barctx);

you need to adjust your function calls accordingly in your application. While the function renaming is a simple regex-search-and-replace and the addition of one parameter introduces compiler errors which you can fix one-by-one, the swapping of the middle parameters is tricky because the types are the same. Now if you miss to adjust the order of the arguments in just a few locations,

    -foo(a, in, out);
	+foo_ex(a, in, out, barctx);

you might end up shipping a program that behaves strangely or crashes in rare occasions.

Reply:
OTOH, the same might happen to you if you program new code and the argument lists are inconsistent...

Reply:
@mspncp as I said in the openssl-project posts (where the original discussions occurred) I strongly believe that _ex functions should preserve the existing argument order because of the confusion it can create and the whole stated `we need to be able to drop ENGINE` is not an argument that holds water as leaving an ENGINE typedef and allowing NULL usage (which is already used as a concept) is actually a lot less disruptive to existing applications. 

We should be aiming to minimise the pain levels for the majority of users - that was our explicit objective here - and I have argued (and continue to argue) that we shouldn't be making things unnecessarily more painful without a substantial benefit to the code base.

I would preserve the existing function argument order as the rule to follow. And remember we will end up with ex2,3,4,etc at times. A fair amount of API disruption can be avoided by using the settable parameters approach rather than fixed function arguments too.

Reply:
For how long should the ENGINE type stick around as an argument?  5 years from now, will it make sense to still have it?

I don't think that anyone is seriously suggesting that we drop the ENGINE argument nownownow, but it's sensible to think that this will happen some time in the future.  It's therefore also sensible to include that possibility in the policy, since it's ono our minds (forward thinking).

Reply:
The downside being that we could end up with this..

XXX_ex4(ctx, unused_1, unused, blah, unused, blah1, unused, newparam);

And you could end up answering a lot of questions about why unused_1 doesnt work..

Reply:
> A fair amount of API disruption can be avoided by using the settable parameters approach rather than fixed function arguments too.

*THIS*, I wholeheartedly agree with!
However, we have to recognise that there are places where that is simply not possible in a sensible way...  I don't remember which function it is for the moment, but I clearly remember trying this exercise somewhere, and the result was a worse pile of garbage than to make a function `FOO_ex` with the argument added in.  So while parametrization like you're talking about should be *strongly encouraged*, it should also be done sensibly (the result should be intuitive enough for our users)

Reply:
> And you could end up answering a lot of questions about why unused_1 doesnt work..

... or why it's even there, 5-10 years from now

Reply:
> > > We must not change the order of existing parameters. This would be very confusing.
> 
> > Why is foo_ex(a, in, out, barctx) more confusing than foo_ex(a, out, in, barctx) when clearly the latter is the confusing one in the scope of our updated API where we strive for outputs after inputs?
> 
> I tend towards @mattcaswell's point of view, because I think that reordering the function parameters while extending the parameter list can be error prone and introduce runtime errors. For example, if we swap parameters of the same type while adding others,
> 
> ```
> -int foo(FOO_CTX *a, char *out, char *in);
> +int foo_ex(FOO_CTX *a, char *in, char *out, BAR_CTX *barctx);
> ```
> 
> you need to adjust your function calls accordingly in your application. While the function renaming is a simple regex-search-and-replace and the addition of one parameter introduces compiler errors which you can fix one-by-one, the swapping of the middle parameters is tricky because the types are the same. Now if you miss to adjust the order of the arguments in just a few locations,
> 
> ```
> -foo(a, out, in);
> +foo_ex(a, out, in, barctx);
> ```
> 
> you might end up shipping a program that behaves strangely or crashes in rare occasions.

I am totally in to avoid unnecessary pain for the existing users, but designing our API so that new functions are intentionally inconsistent to provide existing users the ability to adapt their source code via regex instead of evaluating case by case does not seem a compelling enough argument, if the downside is that we fail to attract new users and contributors because our API is guaranteed to stay unfriendly and inconsistent, even for newly added functions.

One of the main reasons why there is active discouragement from experts from using OpenSSL is because of its messy API, and how this too often leads expert and novice alike to misuse it and result in insecure software. 

I believe we have a mission to fix this (in the least disruptive way possible) and that this "do not rearrange existing arguments" will prevent us from doing that.

Reply:
This issue seems to be particularly contentious and  we have a full range of opinions ranging from "we must only append new arguments, and not add, remove or reorder original arguments" to "we should reorder everything in order to conform to the policy". In my view what we have here is a sensible pragmatic middle ground. I realise that not everyone will agree with me.

Reply:
Your _middle ground_ achieves neither of the actual objectives - it turns into a complete mess - neither useful for existing API users to ease migration, nor logical under whatever the new API pattern is. This is one of many contexts where you simply have to pick which objective you want to meet and go with that.

This is one of many contexts where attempting to find a middle ground is simply the wrong approach and if you read back through the comments I think you can see that your current suggestion actually has complaints from **both** directions rather than any level of broad support.

Which objective is more important? That answers the question. And your so called _middle ground_ meets neither.

Reply:
Clearly I disagree with you. While there are complaints from either extremes, I also find comments above supporting the middle ground. IMO this is the most likely place where we will obtain an agreement that can be supported by the majority. It is clear to me that finding a solution that is acceptable to everyone is not possible. The only way to resolve this is to hold a vote. If the vote fails then we will re-evaluate and see if we can find an alternative position.